clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -441,22 +441,28 @@ class radix_sort {
       unsigned_keys[i] = detail::traits<T>::twiddle_in(unsigned_keys[i]);
     }
 
-    while (true) {
+    for (int i = begin_bit; i < end_bit; i += RADIX_BITS) {
       int pass_bits = sycl::min(RADIX_BITS, end_bit - begin_bit);
 
       int ranks[VALUES_PER_THREAD];
       detail::radix_rank<RADIX_BITS, DESCENDING>(_local_memory)
-          .template rank_keys(item, unsigned_keys, ranks, begin_bit, pass_bits);
-      begin_bit += RADIX_BITS;
+          .template rank_keys(item, unsigned_keys, ranks, i, pass_bits);
 
       item.barrier(sycl::access::fence_space::local_space);
 
+      if (is_striped && i >= end_bit) {
```

Comment:
An **optional** suggestion which would improve the code in my opinion would be to just skip this part of the loop for the last iteration and instead of passing a bool `is_striped` flag to the helper, just implement the last exchange specialization in the public functions which call the helper.

Reply:
Lets not worry about this for now.  It complicates the twiddle out and isn't necessarily better.